### ISSUE TITLE
Caching too much

### ISSUE THREAD

**Issue by carsongee** [mitodl/lore#528]
From running a profiler it looks like we are spending 1/4 of our time on cache sets and gets:
![image](https://cloud.githubusercontent.com/assets/596029/9315931/30c94080-4500-11e5-96ff-20bea888074d.png)

Even though only 139 calls are being made to `get_vocabs`:
![image](https://cloud.githubusercontent.com/assets/596029/9315979/6c6dd902-4500-11e5-8442-f8215f56fd3d.png)

From looking at:
https://github.com/mitodl/lore/blob/master/search/search_indexes.py#L71-L75

It looks like this is probably as a result of generating a cache for all LRs and I think this number will only get worse as more courses are stored (this is on a stack with only the one course).

**Comment by ShawnMilo** [comment]
The problem is that if we cache only a single LearningResource in the context of importing a course, we will do database lookups per `LearningResource`, which is very wasteful.

Maybe we can figure out a way to only populate the entire cache during course import without doing something ugly in the code. Setting a Redis (or cache) key with the course ID upon creation of the `Course` object could work.

This is assuming this issue implies normal, non-import use, where someone saves a `LearningResource` for some reason via the browser independent of an import.

**Comment by carsongee** [comment]
Is it more wasteful that doing a cache set and get for each LR in the entire DB for each new resource (about 30k LRs in lore CI right now for example)? If that is true, that is cool, I'll go looking elsewhere, this just ate up 1/4 of the time of an import of the demo course on a fresh blank install, so it jumped out in the profile data, but I can keep looking.

**Comment by ShawnMilo** [comment]
The way it works now, if it goes to get the vocab info for any `LearningResource` during indexing and finds no result in the cache, it fills the cache for that entire course. This is to make course imports faster. There are tests that show caching results in fewer queries.

However, when that's triggered during non-import use, it's clearly very wasteful. I agree it should be fixed, I just think we should continue to cache for the entire course during imports. It should only be triggered (outside of imports) when a `LearningResource` is saved or a taxonomy term is added for a resource.

**Comment by carsongee** [comment]
Looks like 18% of the time was spent saving to the DB as well, and I mispoke on the 1/4, it was half of the cumulative time that was spent in `get_vocabs`, it was a 47s import

**Comment by carsongee** [comment]
It isn't filling the cache for the entire course though right?  It is doing it for the entire database?  It is `LearningResources.objects.all()`, not `LearningResources.objects.filter(course=course)`, is it maybe just a little mixup, or does it need to do `.all()`?

**Comment by ShawnMilo** [comment]
No, it's properly only filling the cache for the course. However, you did find a bug, which I just filed as #529 and will fix today.

**Comment by carsongee** [comment]
So the fix definitely helped for installations where there are a lot of assets in total, but get_vocab was still taking a fair amount time, and from looking at the profile, it looked like the cache was never getting used (it was called 139 times on a course import with 139 LRs.  I added an exception raise to confirm:

``` diff
index 47ee3d7..3577496 100644
--- a/search/search_indexes.py
+++ b/search/search_indexes.py
@@ -62,6 +62,7 @@ def get_vocabs(course_id, resource_id, solo_update=False):
     key = "vocab_cache_{0}".format(resource_id)
     cached = cache.get(key)
     if (solo_update is False) and (cached is not None):
+        raise Exception
         return cached

     # Pre-populate the cache with blank values in case there are no
```

and the exception was never raised, so I commented out the cache priming:

``` diff
index 47ee3d7..d3ca1ac 100644
--- a/search/search_indexes.py
+++ b/search/search_indexes.py
@@ -62,17 +62,18 @@ def get_vocabs(course_id, resource_id, solo_update=False):
     key = "vocab_cache_{0}".format(resource_id)
     cached = cache.get(key)
     if (solo_update is False) and (cached is not None):
+        raise Exception
         return cached

     # Pre-populate the cache with blank values in case there are no
     # terms for that LearningResource. Otherwise, looking up the vocabularies
     # for that resource will refill the cache for the entire course. If there
     # is already a value, retain it.
-    resource_ids = LearningResource.objects.filter(
-        course__id=course_id).values_list('id', flat=True)
-    for resource_id in resource_ids:
-        rkey = "vocab_cache_{0}".format(resource_id)
-        cache.set(rkey, cache.get(rkey, {}))
+    # resource_ids = LearningResource.objects.filter(
+    #     course__id=course_id).values_list('id', flat=True)
+    # for resource_id in resource_ids:
+    #     rkey = "vocab_cache_{0}".format(resource_id)
+    #     cache.set(rkey, cache.get(rkey, {}))
     value = {}

     term_cache = defaultdict(lambda: defaultdict(list))
```

and the import time for the demo course went from ~40 seconds to ~28 seconds locally, and the resulting profile looks like a pretty even 3 way split now between sql ops for static assets, sql ops for adding LRs, and ES ops for updating the index.

**Comment by carsongee** [comment]
It looks like MechCX went from hours to about 5 minutes locally with that change as well

**Comment by ShawnMilo** [comment]
Did that hours-to-minutes change occur with branch, or you commenting out the caching?

**Comment by carsongee** [comment]
Just that last git diff you see, so commenting out the caching from master as of last night (around 4).

**Comment by carsongee** [comment]
Makes sense, looking at the profile for that large course with the caching code in it is taking 92% of the total import time:
![image](https://cloud.githubusercontent.com/assets/596029/9359597/efff7dce-465f-11e5-9014-9252d69fe75f.png)
Since it is O(n<sup>2</sup>) if the cache always misses (which it does)

**Comment by carsongee** [comment]
Closed by #556
